### PR TITLE
Omniauth shibboleth for CCID-based authenication

### LINKED ISSUE #370: Avalon 6: CCID Authentication
Previous work:

￼Initial commit: 0e1068c
Login and branding: e86bf82
Shib and email: 30d83fc
Email as User.username: b6cd19d
More shib: a21578b

Will likely need to consult with @weiweishi or @murny on how to test this thing without going live.

### PR DESCRIPTION
**Pull request by jefferya**
Fixes #370 

Adds omniauth shibboleth for CCID-based authentication mirroring Avalon 5

Changes proposed in this pull request:

- [x] add omniauth-shibboleth gem
- [x] update `user.rb to work with shibboleth results - [link](https://github.com/ualbertalib/avalon/blob/feature/ccid-auth-370/app/models/user.rb#L90-L111) - inspired from a21578b, b6cd19d & 30d83fc
- [x] update `devise.rb` initializer
- [x] update login link in header to use shibboleth (see [OmniAuth Documentation for `user_shibboleth_omniauth_authorize_path`](https://github.com/plataformatec/devise/wiki/OmniAuth:-Overview), [Avalon 5](https://github.com/ualbertalib/avalon/blob/master/app/views/devise/sessions/new.html.erb#L24-L29) 
 & e86bf82. 
- [x] update changelog.md - add setup requirements regarding shibboleth-sp and notes Sept 26th comment by Weiwei
- [x] retain `users/auth/identity` route

For reference, previous work in Avalon 5:

￼Initial commit: 0e1068c
Login and branding: e86bf82 (not needed - for the most part?)
Shib and email: 30d83fc
Email as User.username: b6cd19d
More shib: a21578b

To test as of 2018-11-27:

- ${SERVER_NAME}/users/auth/shibboleth
  - redirects to: UofA CCID login then one gets returned to the ERA+AV splashpage)
- ${SERVER_NAME}/Shibboleth.sso/Session
  - returns: session details see below (notice missing e-mail)

```
Miscellaneous
Session Expiration (barring inactivity): 375 minute(s)
Client Address: 129.xx.xx.xx
SSO Protocol: urn:oasis:names:tc:SAML:2.0:protocol
Identity Provider: https://${_____IST____}/saml2/idp/metadata.php
Authentication Time: yyyy-mm-ddT16:08:39Z
Authentication Context Class: urn:oasis:names:tc:SAML:2.0:ac:classes:Password
Authentication Context Decl: (none)

Attributes
eduPersonTargetedID: 1 value(s)
givenName: 1 value(s)
sn: 1 value(s)
uid: 1 value(s)
```

Setup [extracted from comment below]:
In the past, we have followed the documentation here, for the Service Provider installation, using RPM on the staging/production environments (and dev environments back then).
https://wiki.shibboleth.net/confluence/display/SHIB2/NativeSPLinuxInstall

Once it's installed. Shibd service needs to be started. And we can generate metadata from shibboleth to send to the IdP and be able to receive responses from them.

References:
- <https://github.com/plataformatec/devise/wiki/OmniAuth:-Overview>
- <https://www.devmynd.com/blog/configuring-rails-app-single-sign-saml-multiple-providers/>
- <https://wiki.shibboleth.net/confluence/display/SHIB2/NativeSPGettingStarted>
- <https://login.ualberta.ca/saml2/idp/metadata.php>

### CODE REVIEW

**Comment by jefferya** [comment]
This WIP PR is an attempt to synthesize the Avalon 5 CCID-based auth and apply into Avalon 6. 

Questions:
1) @weiweishi: unsure how to setup a shibboleth test-bed and there is a note in the original GitHub issue #370 to consult with you?

**Comment by weiweishi** [comment]
@jefferya In the past, we didn't use docker, Shibboleth Service Provider was installed on the application server to handle authentication. I believe we can continue to do the same in dev and UAT but just add some additional complexity to our docker image. 

In the past, we have followed the documentation here, for the Service Provider installation, using RPM on the staging/production environments (and dev environments back then). 
https://wiki.shibboleth.net/confluence/display/SHIB2/NativeSPLinuxInstall

Once it's installed. Shibd service needs to be started. And we can generate metadata from shibboleth to send to the IdP and be able to receive responses from them. 

In the dev environment we used to use: http://www.testshib.org/ for test authentication. 
Here is a list of attributes that are available through UA's SAML authentication:
https://uofaprod.service-now.com/kb_view.do?sysparm_article=KB0012533

And we will need to send our metadata to IST to complete the testing loop with UofA's SAML, and we will need to figure out where would be the best place to do the test. 

Let's grab 15 mins this afternoon to discuss Shibboleth?

File: app/models/user.rb [hunk 1]
```diff
@@ -84,7 +84,30 @@ def self.find_for_lti(auth_hash, signed_in_resource=nil)
...
+    self.find_or_create_by_username_or_email(auth_hash.info.email, auth_hash.info.email)
+  end
+
+  def self.find_for_api(username, email, signed_in_resource=nil)
+    User.find_by_username(username) ||
+    User.find_by_email(email) ||
+    User.create(:username => username, :email => email)
+  end
+
+  def self.find_for_shibboleth(auth_hash, signed_in_resource=nil)
```
Comment (jefferya) [inline app/models/user.rb]: Is this the best approach? For Shibboleth authentication, this is inspired from a21578b, b6cd19d & 30d83fc whereby the e-mail field is used for lookup in the `user` table an if doesn't exist then a new entry is created with the `username` as the e-mail address. This fails if the e-mail field is not part of the Shibboleth idp response.

Is this the best approach?

File: app/views/modules/_header.html.erb [hunk 2]
```diff
@@ -53,10 +53,14 @@ Unless required by applicable law or agreed to in writing, software distributed
           </div><!-- /.btn-group -->
         <% else %>
           <div class="btn-group login_button pull-right">
-            <%= link_to(new_user_session_path, class: "btn btn-default btn-group", title: "Login") do %>
-              <span class="glyphicon glyphicon-log-in" aria-hidden="true"></span>
-              <%= t('ualberta.era_av_sign_in') %>
-            <% end %>
+            <%- Devise.omniauth_providers.each do |provider| %>
```
Comment (jefferya) [inline app/views/modules/_header.html.erb]: I'm having trouble reproducing the Avalon 5 approach to adding the Shibboleth authentication path to the log-in link (See [Avalon 5]https://github.com/ualbertalib/avalon/blob/master/app/views/devise/sessions/new.html.erb#L24-L29) ) & e86bf82. 

Is the Avalon 5 approach better (I can't find variables `devise_mapping` or `resource_class`) or the following approach implemented in the referenced block?

Comment (jefferya) [reply app/views/modules/_header.html.erb]: Remove 'do' and 'if' blocks and use `omniauth_authorize_path(:user, :shibboleth)` instead?

Comment (pgwillia) [reply app/views/modules/_header.html.erb]: `user_{provider}_omniauth_authorize_path` from the [README](https://github.com/plataformatec/devise/wiki/OmniAuth:-Overview) for example
```
<%= link_to "Sign in with Shibboleth", user_shibboleth_omniauth_authorize_path %>
```

**Comment by jefferya** [comment]
@weiweishi 
I think this is ready for code review.

**Pull request merged**